In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes 

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests 

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans # import k-means from clustering stage

#!conda install -c conda-forge folium=0.5.0 --yes 

import folium # map rendering library

print('Libraries imported.')

Libraries imported.


# Data preprocessing

In [2]:
# Get a list of cities of Washington State
tables = pd.read_html("https://en.wikipedia.org/wiki/List_of_cities_and_towns_in_Washington")

In [3]:
# Prepare dataframe of cities of Washington State
df_wa = tables[0]
df_wa = df_wa[['Name']]
df_wa.rename(columns={'Name': 'City'}, inplace=True)
df_wa = df_wa.reset_index(drop=True)
df_wa.columns = ['City']
df_wa['State']='Washington'
df_wa['Latitude']=''
df_wa['Longitude']=''
df_wa

/opt/conda/envs/Python36/lib/python3.6/site-packages/pandas/core/frame.py:4025: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  return super(DataFrame, self).rename(**kwargs)


City       State Latitude Longitude
0             Aberdeen  Washington                   
1       Airway Heights  Washington                   
2               Algona  Washington                   
3            Anacortes  Washington                   
4            Arlington  Washington                   
5               Asotin  Washington                   
6               Auburn  Washington                   
7    Bainbridge Island  Washington                   
8        Battle Ground  Washington                   
9             Bellevue  Washington                   
10          Bellingham  Washington                   
11         Benton City  Washington                   
12              Bingen  Washington                   
13       Black Diamond  Washington                   
14              Blaine  Washington                   
15         Bonney Lake  Washington                   
16             Bothell  Washington                   
17           Bremerton  Washington                   
18            Brewster  Washington                   
19          Bridgeport  Washington                   
20               Brier  Washington                   
21             Buckley  Washington                   
22              Burien  Washington                   
23          Burlington  Washington                   
24               Camas  Washington                   
25           Carnation  Washington                   
26            Cashmere  Washington                   
27         Castle Rock  Washington                   
28           Centralia  Washington                   
29            Chehalis  Washington                   
30              Chelan  Washington                   
31              Cheney  Washington                   
32            Chewelah  Washington                   
33           Clarkston  Washington                   
34            Cle Elum  Washington                   
35          Clyde Hill  Washington                   
36              Colfax  Washington                   
37       College Place  Washington                   
38            Colville  Washington                   
39             Connell  Washington                   
40          Cosmopolis  Washington                   
41           Covington  Washington                   
42           Davenport  Washington                   
43              Dayton  Washington                   
44           Deer Park  Washington                   
45          Des Moines  Washington                   
46              DuPont  Washington                   
47              Duvall  Washington                   
48      East Wenatchee  Washington                   
49            Edgewood  Washington                   
50             Edmonds  Washington                   
51       Electric City  Washington                   
52          Ellensburg  Washington                   
53                Elma  Washington                   
54              Entiat  Washington                   
55            Enumclaw  Washington                   
56             Ephrata  Washington                   
57             Everett  Washington                   
58             Everson  Washington                   
59         Federal Way  Washington                   
60            Ferndale  Washington                   
61                Fife  Washington                   
62            Fircrest  Washington                   
63               Forks  Washington                   
64              George  Washington                   
65          Gig Harbor  Washington                   
66            Gold Bar  Washington                   
67          Goldendale  Washington                   
68        Grand Coulee  Washington                   
69           Grandview  Washington                   
70             Granger  Washington                   
71       Granite Falls  Washington                   
72          Harrington  Washington                   
73             Hoquiam

### Getting latitude and longitude coordinates of each city using MapQuest API

In [4]:
# Add Mapquest credentials to run the following code
MapQ_key = '*******************************'
MapQ_secret = '*******************************'

In [5]:
for index, row in df_wa.iterrows():
    location = row['City'] + "," + row['State']
    url = 'https://www.mapquestapi.com/geocoding/v1/address?key={}&inFormat=kvp&outFormat=json&location={}&thumbMaps=false'.format(
    MapQ_key, 
    location)
    response = requests.get(url).json()
    df_wa.at[index,'Latitude'] = response['results'][0]['locations'][0]['latLng']['lat']
    df_wa.at[index,'Longitude'] = response['results'][0]['locations'][0]['latLng']['lng']
    
df_wa.to_csv('WashingtonCities.csv', sep=',',index=False)

df_wa

City       State Latitude Longitude
0             Aberdeen  Washington  46.9755  -123.816
1       Airway Heights  Washington   47.643  -117.593
2               Algona  Washington  47.2791   -122.25
3            Anacortes  Washington  48.5054  -122.632
4            Arlington  Washington  48.1913  -122.126
5               Asotin  Washington  46.3406  -117.049
6               Auburn  Washington  47.3075  -122.226
7    Bainbridge Island  Washington  47.6431  -122.527
8        Battle Ground  Washington  45.7807  -122.548
9             Bellevue  Washington  47.6137  -122.191
10          Bellingham  Washington  48.7549  -122.478
11         Benton City  Washington  46.2649  -119.488
12              Bingen  Washington  45.7148  -121.466
13       Black Diamond  Washington  47.3087  -122.003
14              Blaine  Washington   48.994  -122.752
15         Bonney Lake  Washington   47.192  -122.181
16             Bothell  Washington  47.7601  -122.205
17           Bremerton  Washington  47.5674  -122.633
18            Brewster  Washington  48.1002  -119.781
19          Bridgeport  Washington  48.0077  -119.671
20               Brier  Washington   47.792  -122.272
21             Buckley  Washington   47.163  -122.026
22              Burien  Washington  47.4667  -122.347
23          Burlington  Washington  48.4769  -122.325
24               Camas  Washington  45.5857  -122.403
25           Carnation  Washington  47.6479  -121.914
26            Cashmere  Washington  47.5174  -120.468
27         Castle Rock  Washington  46.2754  -122.907
28           Centralia  Washington  46.7182  -122.964
29            Chehalis  Washington  46.6619  -122.964
30              Chelan  Washington  47.8474  -120.027
31              Cheney  Washington  47.4955  -117.576
32            Chewelah  Washington  48.2809  -117.724
33           Clarkston  Washington  46.4145  -117.046
34            Cle Elum  Washington  47.1947  -120.936
35          Clyde Hill  Washington   47.632  -122.218
36              Colfax  Washington  46.8838  -117.365
37       College Place  Washington  46.0439  -118.378
38            Colville  Washington  48.5434  -117.906
39             Connell  Washington  46.6648   -118.86
40          Cosmopolis  Washington  46.9559  -123.771
41           Covington  Washington  47.3581  -122.107
42           Davenport  Washington   47.654  -118.158
43              Dayton  Washington  46.3204  -117.979
44           Deer Park  Washington  47.9544   -117.46
45          Des Moines  Washington  47.3916  -122.312
46              DuPont  Washington  47.0974  -122.631
47              Duvall  Washington  47.7401  -121.983
48      East Wenatchee  Washington  47.4162  -120.292
49            Edgewood  Washington    47.25  -122.288
50             Edmonds  Washington  47.8116  -122.377
51       Electric City  Washington  47.9323  -119.038
52          Ellensburg  Washington  46.9996  -120.549
53                Elma  Washington  47.0038  -123.408
54              Entiat  Washington  47.6767  -120.208
55            Enumclaw  Washington  47.2045  -121.992
56             Ephrata  Washington  47.3217  -119.551
57             Everett  Washington  47.9791  -122.196
58             Everson  Washington  48.9111  -122.358
59         Federal Way  Washington  47.3152  -122.335
60            Ferndale  Washington  48.8467  -122.591
61                Fife  Washington   47.239  -122.382
62            Fircrest  Washington  47.2387  -122.515
63               Forks  Washington   47.948  -124.385
64              George  Washington  47.0792  -119.856
65          Gig Harbor  Washington  47.3278  -122.588
66            Gold Bar  Washington  47.8542  -121.694
67          Goldendale  Washington  45.8207  -120.823
68        Grand Coulee  Washington  47.9397  -119.011
69           Grandview  Washington  46.2475  -119.916
70             Granger  Washington  46.3397  -120.187
71       Granite Falls  Washington  48.0834  -121.969
72          Harrington  Washington  47.4788  -118.255
73             Hoquiam

In [6]:
#df = pd.read_csv('WashingtonCity.csv')

In [7]:
# Use geopy library to get the latitude and longitude values of Washington State
address = 'Washington State, USA'

geolocator = Nominatim(user_agent="northwest_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The latitude and longitude of Washington respectively are {}, {}.'.format(latitude, longitude))


The latitude and longitude of Washington respectively are 47.2868352, -120.2126139.


### Visualizing Map of Washington State

In [8]:
# create map of the Washington State using latitude and longitude values
map_wa = folium.Map(location=[latitude, longitude], zoom_start=7)

# add markers to map
for lat, lng, name in zip(df_wa['Latitude'], df_wa['Longitude'], df_wa['City']):
    label = '{}'.format(name)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_wa)  
    
map_wa

### Requesting Top Venues Using Foursquare API

In [9]:
# Use Foursquare developer credentials and add them here
CLIENT_ID = '*******************************************' # your Foursquare ID
CLIENT_SECRET = '*******************************************' # your Foursquare Secret
VERSION = '20200202' # Foursquare API version

In [10]:
# Check
city_latitude = df_wa.loc[0, 'Latitude'] # City latitude value
city_longitude = df_wa.loc[0, 'Longitude'] # City longitude value

city_name = df_wa.loc[0, 'City'] # Name
city_state = df_wa.loc[0, 'State'] # State

print('Latitude and longitude values of {}, {} are: {}, {}.'.format(city_name,
                                                               city_state,
                                                               city_latitude, 
                                                               city_longitude))

Latitude and longitude values of Aberdeen, Washington are: 46.97551, -123.815517.


In [11]:
# Get the top 100 venues within the default city radius
LIMIT = 100 # limit of number of venues returned by Foursquare API

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    city_latitude, 
    city_longitude, 
    LIMIT)

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ed13b8e882fc7001b9ca675'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': '$-$$$$', 'key': 'price'},
    {'name': 'Open now', 'key': 'openNow'}]},
  'suggestedRadius': 10000,
  'headerLocation': 'Aberdeen',
  'headerFullLocation': 'Aberdeen',
  'headerLocationGranularity': 'city',
  'totalResults': 69,
  'suggestedBounds': {'ne': {'lat': 46.98438310725092,
    'lng': -123.7037284668247},
   'sw': {'lat': 46.943570747730774, 'lng': -123.91647086388492}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '52013263498ec644f18a2f77',
       'name': 'Rediviva Restaurant',
       'location': {'address': '118 E Wishkah',
        'lat': 46.97482644098638,
        'lng': -123.81710478759817,
        'labeled

In [12]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [13]:
# Clean the data and structure it as a dataframe
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues = nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name               categories        lat  \
0           Rediviva Restaurant  New American Restaurant  46.974826   
1                    Oceana Spa                      Spa  46.971251   
2     Amore' Italian Restaurant       Italian Restaurant  46.973007   
3             Dutch Bros Coffee                     Café  46.977405   
4  Sucher & Sons Star Wars Shop         Toy / Game Store  46.976254   

          lng  
0 -123.817105  
1 -123.822982  
2 -123.817937  
3 -123.808403  
4 -123.813901

In [14]:
nearby_venues.to_csv('NearbyVenues.csv', sep=',',index=False)

In [15]:
#nearby_venues = pd.read_csv('NearbyVenues.csv')

In [16]:
# Create a function to repeat the same process to all cities
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['City', 
                  'City Latitude', 
                  'City Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [17]:
# Run the function on each city and store in new dataframe
washington_venues = getNearbyVenues(names=df_wa['City'],
                                   latitudes=df_wa['Latitude'],
                                   longitudes=df_wa['Longitude'])

Aberdeen
Airway Heights
Algona
Anacortes
Arlington
Asotin
Auburn
Bainbridge Island
Battle Ground
Bellevue
Bellingham
Benton City
Bingen
Black Diamond
Blaine
Bonney Lake
Bothell
Bremerton
Brewster
Bridgeport
Brier
Buckley
Burien
Burlington
Camas
Carnation
Cashmere
Castle Rock
Centralia
Chehalis
Chelan
Cheney
Chewelah
Clarkston
Cle Elum
Clyde Hill
Colfax
College Place
Colville
Connell
Cosmopolis
Covington
Davenport
Dayton
Deer Park
Des Moines
DuPont
Duvall
East Wenatchee
Edgewood
Edmonds
Electric City
Ellensburg
Elma
Entiat
Enumclaw
Ephrata
Everett
Everson
Federal Way
Ferndale
Fife
Fircrest
Forks
George
Gig Harbor
Gold Bar
Goldendale
Grand Coulee
Grandview
Granger
Granite Falls
Harrington
Hoquiam
Ilwaco
Issaquah
Kahlotus
Kalama
Kelso
Kenmore
Kennewick
Kent
Kettle Falls
Kirkland
Kittitas
La Center
Lacey
Lake Forest Park
Lake Stevens
Lakewood
Langley
Leavenworth
Liberty Lake
Long Beach
Longview
Lynden
Lynnwood
Mabton
Maple Valley
Marysville
Mattawa
McCleary
Medical Lake
Medina
Mercer Islan

In [18]:
washington_venues.to_csv('WashingtonVenues.csv', sep=',',index=False)

In [19]:
# washington_venues = pd.read_csv('WashingtonVenues.csv')

In [20]:
print(washington_venues.shape)
washington_venues.head()

(2740, 7)


City  City Latitude  City Longitude                           Venue  \
0  Aberdeen       46.97551     -123.815517             Rediviva Restaurant   
1  Aberdeen       46.97551     -123.815517    Sucher & Sons Star Wars Shop   
2  Aberdeen       46.97551     -123.815517           Billy's Bar and Grill   
3  Aberdeen       46.97551     -123.815517       Amore' Italian Restaurant   
4  Aberdeen       46.97551     -123.815517  Breakwater Seafood And Chowder   

   Venue Latitude  Venue Longitude           Venue Category  
0       46.974826      -123.817105  New American Restaurant  
1       46.976254      -123.813901         Toy / Game Store  
2       46.975505      -123.813540                      Bar  
3       46.973007      -123.817937       Italian Restaurant  
4       46.975644      -123.811814       Seafood Restaurant

In [21]:
# Tally up the total venues per city
grouped = washington_venues.groupby('City').count()
print('Original count of cities: {}'.format(len(grouped.index)))

# Drop cities with inadequate amount of venue data (they skew the clustering results)
grouped = grouped[grouped.Venue < 10]
list(grouped.index.values)
northwest_venues = washington_venues[~washington_venues['City'].isin(list(grouped.index.values))]
grouped2 = northwest_venues.groupby('City').count()

print('Count of cities with more than 10 venues: {}'.format(len(grouped2.index)))

Original count of cities: 208
Count of cities with more than 10 venues: 82


In [22]:
print('There are {} unique categories.'.format(len(washington_venues['Venue Category'].unique())))

There are 301 unique categories.


### One Hot Encoding

In [23]:
washington_onehot = pd.get_dummies(washington_venues[['Venue Category']], prefix="", prefix_sep="")

# Add city column back to dataframe
washington_onehot['City'] = washington_venues['City'] 

# move city column to the first column
fixed_columns = [washington_onehot.columns[-1]] + list(washington_onehot.columns[:-1])
washington_onehot = washington_onehot[fixed_columns]

# Check size of new dataframe
washington_onehot.shape

(2740, 301)

In [24]:
washington_grouped = washington_onehot.groupby('City').mean().reset_index()
washington_grouped

City  Yoga Studio       ATM  Accessories Store  \
0             Aberdeen     0.000000  0.000000           0.000000   
1       Airway Heights     0.000000  0.000000           0.000000   
2               Algona     0.000000  0.000000           0.000000   
3            Anacortes     0.000000  0.000000           0.000000   
4            Arlington     0.000000  0.000000           0.000000   
5               Asotin     0.000000  0.000000           0.000000   
6               Auburn     0.000000  0.000000           0.000000   
7    Bainbridge Island     0.000000  0.000000           0.000000   
8        Battle Ground     0.000000  0.000000           0.000000   
9             Bellevue     0.000000  0.000000           0.017241   
10          Bellingham     0.000000  0.000000           0.000000   
11         Benton City     0.000000  0.000000           0.000000   
12              Bingen     0.000000  0.000000           0.000000   
13       Black Diamond     0.000000  0.000000           0.000000   
14              Blaine     0.000000  0.047619           0.000000   
15         Bonney Lake     0.000000  0.000000           0.000000   
16             Bothell     0.000000  0.000000           0.000000   
17           Bremerton     0.000000  0.000000           0.000000   
18            Brewster     0.000000  0.000000           0.000000   
19          Bridgeport     0.000000  0.000000           0.000000   
20               Brier     0.000000  0.000000           0.000000   
21             Buckley     0.000000  0.000000           0.000000   
22              Burien     0.024390  0.000000           0.000000   
23          Burlington     0.000000  0.000000           0.000000   
24               Camas     0.000000  0.035714           0.000000   
25           Carnation     0.000000  0.000000           0.000000   
26            Cashmere     0.000000  0.000000           0.000000   
27         Castle Rock     0.000000  0.000000           0.000000   
28           Centralia     0.000000  0.000000           0.000000   
29            Chehalis     0.000000  0.000000           0.000000   
30              Chelan     0.000000  0.000000           0.000000   
31              Cheney     0.000000  0.000000           0.000000   
32            Chewelah     0.000000  0.000000           0.000000   
33           Clarkston     0.000000  0.000000           0.000000   
34            Cle Elum     0.000000  0.000000           0.000000   
35              Colfax     0.000000  0.000000           0.000000   
36       College Place     0.000000  0.000000           0.000000   
37            Colville     0.000000  0.000000           0.000000   
38             Connell     0.000000  0.000000           0.000000   
39          Cosmopolis     0.000000  0.000000           0.000000   
40           Covington     0.027027  0.027027           0.000000   
41           Davenport     0.000000  0.000000           0.000000   
42              Dayton     0.000000  0.000000           0.000000   
43          Des Moines     0.000000  0.000000           0.000000   
44              DuPont     0.000000  0.000000           0.000000   
45              Duvall     0.000000  0.000000           0.000000   
46      East Wenatchee     0.000000  0.000000           0.000000   
47            Edgewood     0.000000  0.000000           0.000000   
48             Edmonds     0.000000  0.021739           0.000000   
49       Electric City     0.000000  0.000000           0.000000   
50          Ellensburg     0.000000  0.035714           0.000000   
51                Elma     0.000000  0.000000           0.000000   
52              Entiat     0.000000  0.000000           0.000000   
53            Enumclaw     0.000000  0.000000           0.000000   
54             Ephrata     0.000000  0.058824           0.000000   
55             Everett     0.000000  0.000000           0.000000   
56             Everson     0.000000  0.000000           0.000000   
57         Federal Way     0.000000  0.000000           0.000000   
58    

In [25]:
# New dataframe size
washington_grouped.shape

(208, 301)

### Top 5 Each City Avenues

In [26]:
num_top_venues = 5

for city in washington_grouped['City']:
    print("----"+city+"----")
    temp = washington_grouped[washington_grouped['City'] == city].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Aberdeen----
                  venue  freq
0  Fast Food Restaurant  0.11
1           Coffee Shop  0.11
2    Italian Restaurant  0.06
3                 Hotel  0.06
4                   Bar  0.06


----Airway Heights----
                 venue  freq
0          Video Store  0.09
1   Chinese Restaurant  0.09
2        Grocery Store  0.09
3                 Park  0.09
4  Japanese Restaurant  0.09


----Algona----
               venue  freq
0               Food  0.17
1      Moving Target  0.17
2                Bar  0.17
3  Convenience Store  0.17
4        Pizza Place  0.17


----Anacortes----
               venue  freq
0       Home Service  0.33
1  Martial Arts Dojo  0.33
2       Soccer Field  0.33
3        Yoga Studio  0.00
4  Other Repair Shop  0.00


----Arlington----
                     venue  freq
0          Automotive Shop  0.17
1      American Restaurant  0.17
2      Arts & Crafts Store  0.17
3  New American Restaurant  0.17
4           Farmers Market  0.17


----Asotin----
        

In [27]:
# Function to sort venues in decscending order
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [28]:
# Create a dataframe with top 5 venues for each city
num_top_venues = 5

indicators = ['st', 'nd', 'rd']

# Create columns according to number of top venues
columns = ['City']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# Create a new dataframe
cities_venues_sorted = pd.DataFrame(columns=columns)
cities_venues_sorted['City'] = washington_grouped['City']

for ind in np.arange(washington_grouped.shape[0]):
    cities_venues_sorted.iloc[ind, 1:] = return_most_common_venues(washington_grouped.iloc[ind, :], num_top_venues)

cities_venues_sorted.head()

City 1st Most Common Venue 2nd Most Common Venue  \
0        Aberdeen  Fast Food Restaurant           Coffee Shop   
1  Airway Heights           Pizza Place           Video Store   
2          Algona      Business Service           Pizza Place   
3       Anacortes     Martial Arts Dojo          Soccer Field   
4       Arlington        Farmers Market                 Diner   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
0           Pizza Place    Italian Restaurant           Music Venue  
1                  Café                   Bar       Thai Restaurant  
2     Convenience Store                  Food         Moving Target  
3          Home Service         Women's Store                  Farm  
4   American Restaurant   Arts & Crafts Store       Automotive Shop

### KMeans Clustering

In [29]:
# Run K-means to break up into clusters
kclusters = 6

washington_grouped_clustering = washington_grouped.drop('City', 1)

# Run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(washington_grouped_clustering)

# Check cluster labels generated for each row in the dataframe
kmeans.labels_[0:300]

array([3, 3, 3, 3, 3, 4, 3, 3, 3, 3, 3, 3, 3, 3, 3, 4, 3, 3, 3, 3, 3, 3,
       3, 3, 3, 3, 3, 3, 3, 3, 3, 0, 1, 3, 3, 3, 0, 3, 0, 3, 3, 3, 3, 3,
       0, 3, 3, 1, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 0, 3, 3, 0,
       0, 2, 0, 3, 0, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
       3, 3, 0, 3, 3, 3, 3, 5, 3, 3, 3, 3, 3, 0, 3, 3, 3, 3, 3, 3, 3, 3,
       3, 3, 3, 3, 3, 3, 3, 3, 0, 3, 0, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 5,
       3, 3, 0, 0, 3, 3, 3, 3, 3, 3, 3, 3, 3, 0, 3, 3, 3, 3, 3, 0, 3, 3,
       3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 1, 3, 4, 3, 3, 0, 3, 3,
       3, 3, 3, 3, 3, 3, 3, 0, 3, 3, 3, 3, 4, 3, 1, 3, 3, 3, 3, 3, 3, 3,
       3, 3, 3, 3, 3, 3, 3, 3, 3, 3], dtype=int32)

In [30]:
cities_venues_sorted.columns

Index(['City', '1st Most Common Venue', '2nd Most Common Venue',
       '3rd Most Common Venue', '4th Most Common Venue',
       '5th Most Common Venue'],
      dtype='object')

In [31]:
cities_venues_sorted

City  1st Most Common Venue       2nd Most Common Venue  \
0             Aberdeen   Fast Food Restaurant                 Coffee Shop   
1       Airway Heights            Pizza Place                 Video Store   
2               Algona       Business Service                 Pizza Place   
3            Anacortes      Martial Arts Dojo                Soccer Field   
4            Arlington         Farmers Market                       Diner   
5               Asotin      Convenience Store               Women's Store   
6               Auburn            Coffee Shop                        Café   
7    Bainbridge Island   Gym / Fitness Center                      Bakery   
8        Battle Ground            Pizza Place                 Coffee Shop   
9             Bellevue            Coffee Shop                       Hotel   
10          Bellingham            Coffee Shop                        Café   
11         Benton City        Automotive Shop                       Diner   
12              Bingen     Mexican Restaurant                 Coffee Shop   
13       Black Diamond              Bookstore                        Food   
14              Blaine                  Diner                 Coffee Shop   
15         Bonney Lake                    Bar               Women's Store   
16             Bothell            Coffee Shop          Mexican Restaurant   
17           Bremerton            Coffee Shop            Asian Restaurant   
18            Brewster        Automotive Shop        Fast Food Restaurant   
19          Bridgeport               Dive Bar                  Food Truck   
20               Brier            Bus Station                   Gift Shop   
21             Buckley           Burger Joint              Sandwich Place   
22              Burien     Mexican Restaurant                Burger Joint   
23          Burlington         Farmers Market                 Pizza Place   
24               Camas            Coffee Shop          Mexican Restaurant   
25           Carnation            Pizza Place                        Café   
26            Cashmere              Wine Shop                  Hobby Shop   
27         Castle Rock          Grocery Store                    Platform   
28           Centralia    Fried Chicken Joint           Convenience Store   
29            Chehalis            Coffee Shop            Insurance Office   
30              Chelan                   Pool                     RV Park   
31              Cheney       Business Service                        Park   
32            Chewelah            Pizza Place                Home Service   
33           Clarkston             Restaurant                  Sports Bar   
34            Cle Elum                 Bakery         American Restaurant   
35              Colfax               Pharmacy         American Restaurant   
36       College Place                   Park          Photography Studio   
37            Colville    American Restaurant               Grocery Store   
38             Connell            Post Office        Gym / Fitness Center   
39          Cosmopolis     Chinese Restaurant     Health & Beauty Service   
40           Covington         Sandwich Place           Convenience Store   
41           Davenport            Gas Station          Mexican Restaurant   
42              Dayton                  Hotel                  Restaurant   
43          Des Moines              Locksmith                        Food   
44              DuPont         History Museum            Business Service   
45              Duvall            Coffee Shop                 Pizza Place   
46      East Wenatchee            Coffee Shop           Convenience Store   
47            Edgewood            Coffee Shop                         Gym   
48             Edmonds                   Café                         Bar   
49       Electric City              Gastropub                         Pub   
50          Ellensburg          Grocery Store          Italian Restaurant   
51            

In [32]:
# Create dataframe that includes the cluster and top 5 venues

# Add clustering labels
cities_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

washington_merged = df_wa

# Merge washington_grouped with northwest_data to add latitude/longitude for each city
washington_merged = washington_merged.join(cities_venues_sorted.set_index('City'), on='City')

# Drop cities with no venue data
washington_merged = washington_merged.dropna()

washington_merged

City       State Latitude Longitude  Cluster Labels  \
0             Aberdeen  Washington  46.9755  -123.816             3.0   
1       Airway Heights  Washington   47.643  -117.593             3.0   
2               Algona  Washington  47.2791   -122.25             3.0   
3            Anacortes  Washington  48.5054  -122.632             3.0   
4            Arlington  Washington  48.1913  -122.126             3.0   
5               Asotin  Washington  46.3406  -117.049             4.0   
6               Auburn  Washington  47.3075  -122.226             3.0   
7    Bainbridge Island  Washington  47.6431  -122.527             3.0   
8        Battle Ground  Washington  45.7807  -122.548             3.0   
9             Bellevue  Washington  47.6137  -122.191             3.0   
10          Bellingham  Washington  48.7549  -122.478             3.0   
11         Benton City  Washington  46.2649  -119.488             3.0   
12              Bingen  Washington  45.7148  -121.466             3.0   
13       Black Diamond  Washington  47.3087  -122.003             3.0   
14              Blaine  Washington   48.994  -122.752             3.0   
15         Bonney Lake  Washington   47.192  -122.181             4.0   
16             Bothell  Washington  47.7601  -122.205             3.0   
17           Bremerton  Washington  47.5674  -122.633             3.0   
18            Brewster  Washington  48.1002  -119.781             3.0   
19          Bridgeport  Washington  48.0077  -119.671             3.0   
20               Brier  Washington   47.792  -122.272             3.0   
21             Buckley  Washington   47.163  -122.026             3.0   
22              Burien  Washington  47.4667  -122.347             3.0   
23          Burlington  Washington  48.4769  -122.325             3.0   
24               Camas  Washington  45.5857  -122.403             3.0   
25           Carnation  Washington  47.6479  -121.914             3.0   
26            Cashmere  Washington  47.5174  -120.468             3.0   
27         Castle Rock  Washington  46.2754  -122.907             3.0   
28           Centralia  Washington  46.7182  -122.964             3.0   
29            Chehalis  Washington  46.6619  -122.964             3.0   
30              Chelan  Washington  47.8474  -120.027             3.0   
31              Cheney  Washington  47.4955  -117.576             0.0   
32            Chewelah  Washington  48.2809  -117.724             1.0   
33           Clarkston  Washington  46.4145  -117.046             3.0   
34            Cle Elum  Washington  47.1947  -120.936             3.0   
36              Colfax  Washington  46.8838  -117.365             3.0   
37       College Place  Washington  46.0439  -118.378             0.0   
38            Colville  Washington  48.5434  -117.906             3.0   
39             Connell  Washington  46.6648   -118.86             0.0   
40          Cosmopolis  Washington  46.9559  -123.771             3.0   
41           Covington  Washington  47.3581  -122.107             3.0   
42           Davenport  Washington   47.654  -118.158             3.0   
43              Dayton  Washington  46.3204  -117.979             3.0   
45          Des Moines  Washington  47.3916  -122.312             3.0   
46              DuPont  Washington  47.0974  -122.631             0.0   
47              Duvall  Washington  47.7401  -121.983             3.0   
48      East Wenatchee  Washington  47.4162  -120.292             3.0   
49            Edgewood  Washington    47.25  -122.288             1.0   
50             Edmonds  Washington  47.8116  -122.377             3.0   
51       Electric City  Washington  47.9323  -119.038             3.0   
52          Ellensburg  Washington  46.9996  -120.549             3.0   
53                Elma  Washington  47.0038  -123.408             3.0   
54              Entiat  Washington  47.6767  -120.208             3.0   
55            Enumclaw  Washington  47.2045  -121.992             3.0   
56 

### Mapping Result of KMeans Cluster

In [33]:
# Create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=7)

# Set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# Add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(washington_merged['Latitude'], washington_merged['Longitude'], washington_merged['City'], washington_merged['Cluster Labels']):
    cluster = int(cluster)
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Project Result

Cluster 0: It has 19 cities with mainly parks and farms avenue. So I think these cities have many wide open areas and suitable for people who love nature. 

In [34]:
# Cluster 0
washington_merged.loc[washington_merged['Cluster Labels'] == 0, washington_merged.columns[[0] + [1] + list(range(5, washington_merged.shape[1]))]]

City       State 1st Most Common Venue    2nd Most Common Venue  \
31          Cheney  Washington      Business Service                     Park   
37   College Place  Washington                  Park       Photography Studio   
39         Connell  Washington           Post Office     Gym / Fitness Center   
46          DuPont  Washington        History Museum         Business Service   
64          George  Washington            Food Truck  State / Provincial Park   
67      Goldendale  Washington            Restaurant                  Brewery   
68    Grand Coulee  Washington        Ice Cream Shop              Pizza Place   
70         Granger  Washington                  Park            Women's Store   
72      Harrington  Washington     Convenience Store                     Park   
92    Liberty Lake  Washington                  Pool              Art Gallery   
103         Medina  Washington            Playground                     Park   
120        Newport  Washington     Food & Drink Shop                     Park   
122  Normandy Park  Washington                  Pool         Sculpture Garden   
137        Pateros  Washington                   Inn                   Bakery   
138        Pomeroy  Washington         Grocery Store                     Park   
148        Rainier  Washington                  Park             Dessert Shop   
154     Ridgefield  Washington               Brewery       Mexican Restaurant   
176        Sprague  Washington                 Motel                     Park   
186         Tieton  Washington   Sporting Goods Shop                     Park   

          3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
31                          Bar      Volleyball Court         Women's Store  
37   Construction & Landscaping           Fish Market                  Farm  
39                  Pizza Place         Movie Theater                  Park  
46                         Park          Burger Joint                  Farm  
64               Sandwich Place                  Park       Fireworks Store  
67                  Pizza Place                   Bar               RV Park  
68                         Park         Grocery Store     Fondue Restaurant  
70            Fish & Chips Shop    Falafel Restaurant                  Farm  
72                Women's Store           Fish Market                  Farm  
92                         Park      Business Service                  Food  
103                     Dog Run         Women's Store     Fish & Chips Shop  
120                      Bakery         Women's Store           Fish Market  
122        Gym / Fitness Center                  Park                 Beach  
137                        Park   American Restaurant         Women's Store  
138                 Golf Course   American Restaurant     Fish & Chips Shop  
148            Asian Restaurant           Gas Station         Women's Store  
154                        Park   American Restaurant         Women's Store  
176         American Restaurant         Women's Store           Fish Market  
186                 Art Gallery           Event Space                  Farm

Cluster 1: There are only four cities listed in cluster 1

In [35]:
# Cluster 1
washington_merged.loc[washington_merged['Cluster Labels'] == 1, washington_merged.columns[[0] + [1] + list(range(5, washington_merged.shape[1]))]]

City       State 1st Most Common Venue 2nd Most Common Venue  \
32           Chewelah  Washington           Pizza Place          Home Service   
49           Edgewood  Washington           Coffee Shop                   Gym   
171         Soap Lake  Washington           Coffee Shop         Women's Store   
193  University Place  Washington                   Spa           Coffee Shop   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
32          Women's Store           Fish Market    Falafel Restaurant  
49      Convenience Store           Pizza Place          Home Service  
171          Fishing Spot                  Farm        Farmers Market  
193         Women's Store           Fish Market                  Farm

Cluster 2: Based on the KMeans Clustering method, Grandview City is the most unique place in the Washington State

In [36]:
# Cluster 2
washington_merged.loc[washington_merged['Cluster Labels'] == 2, washington_merged.columns[[0] + [1] + list(range(5, washington_merged.shape[1]))]]

City       State 1st Most Common Venue 2nd Most Common Venue  \
69  Grandview  Washington        Nightlife Spot         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
69           Event Space                  Farm        Farmers Market

Cluster 3: This cluster has many cities on the list. The most striking feature is that Seattle belongs to this cluster. So, I think these cities on the list have many similarities to Seattle and might be people's alternative to find home where the city are less populated than Seattle but have Seattle vibes. More over, the cities on this cluster have many entertainment avenues as the top five avenues such bar, restaurant, cafe, coffe shop, and many more which is good for bussiness.   

In [37]:
# Cluster 3
washington_merged.loc[washington_merged['Cluster Labels'] == 3, washington_merged.columns[[0] + [1] + list(range(5, washington_merged.shape[1]))]]

City       State  1st Most Common Venue  \
0             Aberdeen  Washington   Fast Food Restaurant   
1       Airway Heights  Washington            Pizza Place   
2               Algona  Washington       Business Service   
3            Anacortes  Washington      Martial Arts Dojo   
4            Arlington  Washington         Farmers Market   
6               Auburn  Washington            Coffee Shop   
7    Bainbridge Island  Washington   Gym / Fitness Center   
8        Battle Ground  Washington            Pizza Place   
9             Bellevue  Washington            Coffee Shop   
10          Bellingham  Washington            Coffee Shop   
11         Benton City  Washington        Automotive Shop   
12              Bingen  Washington     Mexican Restaurant   
13       Black Diamond  Washington              Bookstore   
14              Blaine  Washington                  Diner   
16             Bothell  Washington            Coffee Shop   
17           Bremerton  Washington            Coffee Shop   
18            Brewster  Washington        Automotive Shop   
19          Bridgeport  Washington               Dive Bar   
20               Brier  Washington            Bus Station   
21             Buckley  Washington           Burger Joint   
22              Burien  Washington     Mexican Restaurant   
23          Burlington  Washington         Farmers Market   
24               Camas  Washington            Coffee Shop   
25           Carnation  Washington            Pizza Place   
26            Cashmere  Washington              Wine Shop   
27         Castle Rock  Washington          Grocery Store   
28           Centralia  Washington    Fried Chicken Joint   
29            Chehalis  Washington            Coffee Shop   
30              Chelan  Washington                   Pool   
33           Clarkston  Washington             Restaurant   
34            Cle Elum  Washington                 Bakery   
36              Colfax  Washington               Pharmacy   
38            Colville  Washington    American Restaurant   
40          Cosmopolis  Washington     Chinese Restaurant   
41           Covington  Washington         Sandwich Place   
42           Davenport  Washington            Gas Station   
43              Dayton  Washington                  Hotel   
45          Des Moines  Washington              Locksmith   
47              Duvall  Washington            Coffee Shop   
48      East Wenatchee  Washington            Coffee Shop   
50             Edmonds  Washington                   Café   
51       Electric City  Washington              Gastropub   
52          Ellensburg  Washington          Grocery Store   
53                Elma  Washington                   Bank   
54              Entiat  Washington          Grocery Store   
55            Enumclaw  Washington            Pizza Place   
56             Ephrata  Washington                   Food   
57             Everett  Washington    American Restaurant   
58             Everson  Washington               Ski Area   
59         Federal Way  Washington        Assisted Living   
60            Ferndale  Washington                    Bar   
61                Fife  Washington          Boat or Ferry   
62            Fircrest  Washington                   Café   
63               Forks  Washington            Coffee Shop   
65          Gig Harbor  Washington              BBQ Joint   
66            Gold Bar  Washington         Breakfast Spot   
71       Granite Falls  Washington            Coffee Shop   
73             Hoquiam  Washington     Mexican Restaurant   
74              Ilwaco  Washington     Seafood Restaurant   
75            Issaquah  Washington            Pizza Place   
76            Kahlotus  Washington                   Café   
77              Kalama  Washington             Kids Store   
78               Kelso  Washington            Bus Station   
79             Kenmore  Washington            Pizza Place   
80           Kennewick  Washington                   

Cluster 4: It has only four cities listed on this cluster. The top avenues here are fishing spot and farmers market. I think this cluster has similar condition with cluster 0. 

In [38]:
# Cluster 4
washington_merged.loc[washington_merged['Cluster Labels'] == 4, washington_merged.columns[[0] + [1] + list(range(5, washington_merged.shape[1]))]]

City       State 1st Most Common Venue 2nd Most Common Venue  \
5         Asotin  Washington     Convenience Store         Women's Store   
15   Bonney Lake  Washington                   Bar         Women's Store   
173      Spangle  Washington                   Bar   American Restaurant   
191     Tumwater  Washington     Convenience Store                   Bar   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
5            Fishing Spot                  Farm        Farmers Market  
15           Fishing Spot        Farmers Market  Fast Food Restaurant  
173         Women's Store          Fishing Spot        Farmers Market  
191         Women's Store          Fishing Spot        Farmers Market

Cluster 5: This cluster has unique avenues on the top 5 avenues such as candy store and falafel restaurant making them unique in this dataset

In [39]:
# Cluster 5
washington_merged.loc[washington_merged['Cluster Labels'] == 5, washington_merged.columns[[0] + [1] + list(range(5, washington_merged.shape[1]))]]

City       State 1st Most Common Venue 2nd Most Common Venue  \
97    Mabton  Washington      Business Service         Women's Store   
134  Pacific  Washington      Business Service                   Gym   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
97     Falafel Restaurant        Farmers Market  Fast Food Restaurant  
134           Candy Store         Women's Store        Farmers Market

## Conclusion

We first webscraped data from Wikipedia to get list of cities in Washington State. There were 212 cities on the listthat we needed to find their location data which were latitude and longitude coordinates of each city. To get those information, we used MapQuest API. After that, we used FourSquare API to get FourSquare query of venues, there were 2740 venues in the Washington State including 301 unique category venues. To solve the probrlem, we used KMeans clustering in order to group the cities into 6 clusters. However there are only four data that are reliable. This project needs to be improved whether using more reliable dataset or clustering method. Based on this project, we sure can answer the main problem stated in the beginning of this project that is "are there nearby city around Washington state that similar to Seattle to live in?". And the answer are listed on cluster 3.